paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc

```diff
@@ -0,0 +1,497 @@
+    const std::unordered_map<std::string, Node*>& tensor_nodes,
+    std::unordered_set<std::string>* free_existing_tensors,
+    const std::unordered_map<std::string, int>& space_table,
+    const std::vector<std::unordered_set<std::string>>& var_clusters,
+    std::string* tensor2use) {
+  std::pair<std::string, int> best_fit;
+  best_fit.second = std::numeric_limits<int>::max();
+  VLOG(3) << "Split Tensors to " << var_clusters.size() << " clusters";
+
+  // find the cluster this var belongs to.
+  const std::unordered_set<std::string>* cluster = nullptr;
+  for (const auto& c : var_clusters) {
+    if (c.count(tensor)) {
+      cluster = &c;
+      break;
+    }
+  }
+  PADDLE_ENFORCE_NOT_NULL(
+      cluster,
+      "something wrong in memory optimization, the variable cluster phase.");
+
+  for (auto& candidate : *free_existing_tensors) {
+    if (!space_table.count(candidate)) continue;
+    int space = space_table.at(candidate);
+    int space_diff = space - space_required;
+    if (space_diff >= 0 && space_diff < best_fit.second &&
+        cluster->count(candidate)) {
+      best_fit.first = candidate;
+      best_fit.second = space_diff;
+    }
+  }
+
+  if (best_fit.second < std::numeric_limits<int>::max()) {
+    *tensor2use = best_fit.first;
+    return true;
+  }
+  return false;
+}
+
+// Allocate new tensor instead of reusing the existing one.
+void AllocateNewTensor(
+    const std::string& name, int space_required,
+    const std::unordered_map<std::string, Node*>& tensor_nodes,
+    std::unordered_set<std::string>* free_existing_tensors,
+    std::unordered_map<std::string, int>* space_table,
+    std::unordered_map<std::string, std::string>* reuse_table) {
+  // The newly born tensor is free to be used.
+  free_existing_tensors->insert(name);
+  // Register the space it has.
+  space_table->emplace(name, space_required);
+  // The allocated new tensor use the memory of itself.
+  (*reuse_table)[name] = name;
+}
+
+// Free a tensor and make it resuable.
+// @tensor: the tensor to free.
+// @free_existing_tensors: the free and allocated tensors.
+// @reuse_table: a map from a fake tensor to the existing allocated tensor.
+void FreeATensor(const std::string& tensor,
+                 std::unordered_set<std::string>* free_existing_tensors,
+                 std::unordered_map<std::string, std::string>* reuse_table) {
+  if (tensor == "feed" || tensor == "fetch") return;
+  // the really allocated tensor.
+  const auto& free_tensor = reuse_table->at(tensor);
+
+  free_existing_tensors->insert(free_tensor);
+}
+
+// Reuse a free existing tensor.
+void ReuseATensor(const std::string& tensor, const std::string& tensor2reuse,
+                  std::unordered_set<std::string>* free_existing_tensors,
+                  std::unordered_map<std::string, std::string>* reuse_table) {
+  auto it = free_existing_tensors->find(tensor2reuse);
+  PADDLE_ENFORCE(it != free_existing_tensors->end());
+  free_existing_tensors->erase(it);
```

Comment:
Why delete it from free_existing_tensors ?

Reply:
the `free_existing_tensors` stores all the existing tensor that has been allocated memory and the memory is ready for reuse.
If an existing tensor is already reused and not free yet, it should be removed from this container variable.